src/resources/ble_darwin.mm

```diff
@@ -0,0 +1,1297 @@
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(discover_services) {
+  ARGS(BLERemoteDeviceResource, device, Array, raw_service_uuids);
+
+  Error* err;
+  NSArray<CBUUID*>* service_uuids = ns_uuid_array_from_array_of_strings(process, raw_service_uuids, &err);
+  if (err) return err;
+  [device->peripheral() discoverServices:service_uuids];
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(discover_services_result) {
+  ARGS(BLERemoteDeviceResource, device);
+
+  NSArray<CBService*>* services = [device->peripheral() services];
+  int count = 0;
+  BLEServiceResource* service_resources[([services count])];
+  for (int i = 0; i < [services count]; i++) {
+    BLEServiceResource* service_resource = device->get_or_create_service_resource(services[i], true);
+    if (service_resource->is_returned()) continue;
+    service_resources[count++] = service_resource;
+  }
+
+  Array* array = process->object_heap()->allocate_array(count, process->program()->null_object());
+  if (array == null) ALLOCATION_FAILED;
+
+  for (int i = 0; i < count; i++) {
+    BLEServiceResource* service_resource = service_resources[i];
+
+    String* uuid_str = process->allocate_string([[services[i].UUID UUIDString] UTF8String]);
+    if (uuid_str == null) ALLOCATION_FAILED;
+    
+    Array* service_info = process->object_heap()->allocate_array(2, process->program()->null_object());
+    if (service_info == null) ALLOCATION_FAILED;
+
+    ByteArray* proxy = process->object_heap()->allocate_proxy();
+    if (proxy == null) ALLOCATION_FAILED;
+    proxy->set_external_address(service_resource);
+
+    service_info->at_put(0, uuid_str);
+    service_info->at_put(1, proxy);
+    array->at_put(i, service_info);
+  }
+
+  // Second loop to mark resources as returned, this way there is no need to do clean up on ALLOCATION_FAILED
+  for (int i = 0; i < count; i++) service_resources[i]->set_returned(true);
+
+  return array;
+}
+
+PRIMITIVE(discover_characteristics) {
+  ARGS(BLEServiceResource, service, Array, raw_characteristics_uuids);
+
+  if (!service->device()) INVALID_ARGUMENT;
+
+  Error* err;
+  NSArray<CBUUID*>* characteristics_uuids = ns_uuid_array_from_array_of_strings(process, raw_characteristics_uuids,
+                                                                                &err);
+  if (err) return err;
+
+  [service->device()->peripheral() discoverCharacteristics:characteristics_uuids forService:service->service()];
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(discover_characteristics_result) {
+  ARGS(BLEServiceResource, service);
+
+  NSArray<CBCharacteristic*>* characteristics = [service->service() characteristics];
+
+  int count = 0;
+  BLECharacteristicResource* characteristic_resources[([characteristics count])];
+  for (int i = 0; i < [characteristics count]; i++) {
+    BLECharacteristicResource* characteristic_resource
+        = service->get_or_create_characteristic_resource(characteristics[i],true);
```

Comment:
unless this is convention in darwin code.
```suggestion
        = service->get_or_create_characteristic_resource(characteristics[i], true);
```

Reply:
It is not. The problem is my editor that does some "clever" annotation, so I dont always catch these missing spaces